core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -136,6 +146,16 @@ public RecoveryResponse recoverToTarget() throws IOException {
                 }
             }
 
+            // engine was just started at the end of phase 1
+            if (shard.state() == IndexShardState.RELOCATED) {
+                /**
+                 * we could have missed replication operations on the target if its engine wasn't started yet after primary relocation
+                 * hand-off completed. The reason is that replication failures are just (silently) ignored if the engine isn't started yet,
+                 * see {@link ReplicationOperation#ignoreReplicaException(Throwable)} and {@link IndexShard#getEngine()}.
+                 */
+                throw new IndexShardRelocatedException(request.shardId());
```

Comment:
do we need to use IndexShardRelocationException? I looked at the other usages of it, and I think we can get of it all together? We can  use IllegalIndexShardStateException ?